### ISSUE TITLE
gnocchi-charm-openstack-services-checks failing

### ISSUE THREAD

**Issue by sudeephb** [canonical/charm-openstack-service-checks#48]
Related to Bug #1826004 reported by Drew Freiberger on 2019-04-23

A fix was released but the issue is still present with TLS and latest gnocchi charm.
Queens Bionic with Contrail
PCB build version cpe-foundation-stable-2.2
Openstack bundle charms 19.07

If nagios is failing for gnocchi with the following message:
```
HTTP CRITICAL - Invalid HTTP response received from host on port 8041: HTTP/1.1 400 Bad Request
```
It is needed to remove and re-establish the relation between `openstack-service-checks` and `keystone`:
```bash
juju remove-relation keystone openstack-service-checks
juju add-relation keystone openstack-service-checks
```

------------------------------------
Imported from Launchpad using lp2gh.

 * date created: 2019-11-14T08:57:37Z
 * owner: [rgrullon](https://launchpad.net/~rgrullon)
 * assignee: [freyes](https://launchpad.net/~freyes)
 
 
 * the launchpad [url](https://bugs.launchpad.net/bugs/1852549)

**Comment by sudeephb** [comment]
(by [afreiberger](https://launchpad.net/~afreiberger))
I think this is happening because services like designate/gnocchi are starting by registering http endpoints and then changing to https endpoints w/out openstack-service-checks getting an update regarding that via relation or update-status checks.

Might suggest adding openstack endpoint list caching and diffing to update-status to automatically re-up this relation, if there's not the ability to react from keystone relation changes based on endpoint updates from other apps.

This happens to both gnocchi and designate.

**Comment by sudeephb** [comment]
(by [freyes](https://launchpad.net/~freyes))
keystone has an interface named identity-notifications, it's used to broadcast changes to endpoints, one of their consumers is ceilometer, openstack-service-checks charms is a good candidate to take advantage of this "bus" too.

**Comment by sudeephb** [comment]
(by [freyes](https://launchpad.net/~freyes))
The following bug in the keystone charm is preventing from getting accurate updates to changes in the catalog: https://bugs.launchpad.net/charm-keystone/+bug/1856419

**Comment by sudeephb** [comment]
(by [freyes](https://launchpad.net/~freyes))
I didn't find any interface that implements the identity-notifications interface, so I implemented one -> https://github.com/freyes/charm-interface-keystone-notifications , it needs to be proposed for inclusion in the "openstack charms" project.

**Comment by sudeephb** [comment]
(by [freyes](https://launchpad.net/~freyes))
A draft of the patch to fix this bug can be found at https://code.launchpad.net/~freyes/charm-openstack-service-checks/+git/charm-openstack-service-checks/+ref/lp1852549

**Comment by sudeephb** [comment]
(by [freyes](https://launchpad.net/~freyes))
Fix available in stable/20.02 (cs:~canonical-bootstack/openstack-service-checks-32)

**Comment by sudeephb** [comment]
(by [freyes](https://launchpad.net/~freyes))
Remember to relate keystone to openstack-service-checks through the identity-notifications (juju add-relation keystone:identity-notifications openstack-service-checks:identity-notifications) which will notify of changes to the keystone catalog.

**Comment by sudeephb** [comment]
(by [oil-ci-bot](https://launchpad.net/~oil-ci-bot))

This bug is fixed with commit cb9b5394 to fce-templates on branch fcb/intermediate/master.
To view that commit see the following URL:
https://git.launchpad.net/fce-templates/commit/?id=cb9b5394
